src/aggregation/agg_result.rs

```diff
@@ -18,55 +17,76 @@ use super::intermediate_agg_result::{
 };
 use super::metric::{SingleMetricResult, Stats};
 use super::Key;
+use crate::TantivyError;
 
 #[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
 /// The final aggegation result.
 pub struct AggregationResults(pub HashMap<String, AggregationResult>);
 
 impl AggregationResults {
+    pub(crate) fn get_value_from_aggregation(
+        &self,
+        name: &str,
+        agg_property: &str,
+    ) -> crate::Result<Option<f64>> {
+        if let Some(agg) = self.0.get(name) {
+            agg.get_value_from_aggregation(name, agg_property)
+        } else {
+            // Validation is be done during request parsing, so we can't reach this state.
+            Err(TantivyError::InternalError(format!(
+                "Can't find aggregation {:?} in sub_aggregations",
+                name
+            )))
+        }
+    }
+
     /// Convert and intermediate result and its aggregation request to the final result
     pub fn from_intermediate_and_req(
         results: IntermediateAggregationResults,
         agg: Aggregations,
-    ) -> Self {
+    ) -> crate::Result<Self> {
         AggregationResults::from_intermediate_and_req_internal(results, &(agg.into()))
     }
+
     /// Convert and intermediate result and its aggregation request to the final result
     ///
     /// Internal function, CollectorAggregations is used instead Aggregations, which is optimized
     /// for internal processing, by splitting metric and buckets into seperate groups.
     pub(crate) fn from_intermediate_and_req_internal(
         results: IntermediateAggregationResults,
         req: &AggregationsInternal,
-    ) -> Self {
-        let mut result = HashMap::default();
-
+    ) -> crate::Result<Self> {
         // Important assumption:
         // When the tree contains buckets/metric, we expect it to have all buckets/metrics from the
         // request
-        if let Some(buckets) = results.buckets {
-            result.extend(buckets.into_iter().zip(req.buckets.values()).map(
-                |((key, bucket), req)| {
-                    (
+        let mut result: HashMap<_, _> = if let Some(buckets) = results.buckets {
```

Comment:
This is very hard to follow.
Can you:
- isolate the two bodies in independent functions?
- add an assert on the stuff being zipped being of the same size.
- use an iterative style maybe?